File: quadratic-client/src/app/gridGL/pixiApp/background.ts
//! This shows the background for the grid. There are two portions of the
//! background: in-bounds and out-of-bounds.

import { Graphics } from 'pixi.js';
import { pixiApp } from './PixiApp';
import { colors } from '@/app/theme/colors';

export class Background extends Graphics {
  update(viewportDirty: boolean) {
    if (!viewportDirty) return;
    this.clear();
    const visibleBounds = pixiApp.viewport.getVisibleBounds();
    if (visibleBounds.left < 0) {
      const right = Math.min(0, visibleBounds.right);
      const bottom = Math.max(0, visibleBounds.bottom);
      this.beginFill(colors.outOfBoundsBackgroundColor);
      this.drawRect(visibleBounds.left, 0, right - visibleBounds.left, bottom);
      this.endFill();
    }
    if (visibleBounds.top < 0) {
      const right = Math.max(0, visibleBounds.right);
      const bottom = Math.min(0, visibleBounds.bottom);
      this.beginFill(colors.outOfBoundsBackgroundColor);
      this.drawRect(visibleBounds.left, visibleBounds.top, right - visibleBounds.left, bottom - visibleBounds.top);
      this.endFill();
    }

    // draw normal area
    this.beginFill(colors.gridBackground);
    const x = Math.max(0, visibleBounds.left);
    const y = Math.max(0, visibleBounds.top);
    this.drawRect(x, y, visibleBounds.right - x, visibleBounds.bottom - y);
    this.endFill();
  }
}

File: quadratic-rust-renderer/wasm/renderer/src/worker/render/background.rs
//! Background rendering (white grid area)

use crate::viewport::Viewport;

/// Get background vertices for rendering
pub fn get_background_vertices(viewport: &Viewport) -> Option<Vec<f32>> {
    let bounds = viewport.visible_bounds();

    if bounds.right <= 0.0 || bounds.bottom <= 0.0 {
        return None;
    }

    let x = bounds.left.max(0.0);
    let y = bounds.top.max(0.0);
    let width = bounds.right - x;
    let height = bounds.bottom - y;

    let x2 = x + width;
    let y2 = y + height;
    let color = [1.0f32, 1.0, 1.0, 1.0];

    // Two triangles forming a rectangle
    Some(vec![
        // Triangle 1
        x,
        y,
        color[0],
        color[1],
        color[2],
        color[3],
        x2,
        y,
        color[0],
        color[1],
        color[2],
        color[3],
        x2,
        y2,
        color[0],
        color[1],
        color[2],
        color[3],
        // Triangle 2
        x,
        y,
        color[0],
        color[1],
        color[2],
        color[3],
        x2,
        y2,
        color[0],
        color[1],
        color[2],
        color[3],
        x,
        y2,
        color[0],
        color[1],
        color[2],
        color[3],
    ])
}
